src/msw/textctrl.cpp

```diff
@@ -2816,7 +2816,7 @@ bool wxTextCtrl::MSWSetCharFormat(const wxTextAttr& style, long start, long end)
         // CHARFORMAT or CHARFORMAT2
         cf.cbSize = sizeof(cf);
     }
-
+    cf.dwMask |= CFM_UNDERLINETYPE;
```

Comment:
It should probably stay here with the dwMask initialization.

Vadim?

Reply:
If I look at for example `cf.dwMask |= CFM_COLOR;`, this mask is only set when the text colour will be changed.
Something similar can be done for the underlined code. Put everything inside an `if (style.HasFontUnderlined())` check. This prevents executing a lot of code when underline is disabled.

Reply:
Except that it also removes the underlining.

Reply:
No, you only need to set the mask when underlining is enabled.
Maybe because currently `HasFontUnderlined` is implemented wrongly it doesnt work for you, it should be reverted back to
```c++
bool HasFontUnderlined() const { return HasFlag(wxTEXT_ATTR_FONT_UNDERLINE); }
```